### ISSUE TITLE
Migration tool should track postdominators of DecoratedTypes to handle hard edges on casts

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/sdk#39555]
When run on the following input:

```dart
class A {
  num get foo => null;
}

void test() {
  A a = A();
  var x = a.foo as int;
  x.isEven;
}
```

The migration tool produces the following output:

```dart
class A {
  num? get foo => null;
}

void test() {
  A a = A();
  var x = a.foo as int?!;
  x.isEven;  
}
```

I believe (and the parser agrees) that the cast `a.foo as int?!` is not valid syntax, since `!` must be applied to a primary expression.  The cast should be parenthesized here.  

A nice bonus would be to instead cast to the non-nullable version of the type, but that's secondary.

cc @stereotype441 

cc @lrhn @eernstg to check my interpretation of the grammar.

**Comment by eernstg** [comment]
I agree on the syntax stuff. Suffix '!' can be put after `<X> <selector>*` where `<X>` can be `<primary>`, `<constructorInvocation>` and a cascade, and `as` is a `<typeCast>` which is much closer to `<expression>`.

**Comment by stereotype441** [comment]
Agreed--this is crazy output and needs to be fixed.  I'm currently reworking the way the migration tool builds its set of fixes (#38472).  Once that is done I can work on this.

**Comment by srawlins** [comment]
Duplicate of https://github.com/dart-lang/sdk/issues/38469

**Comment by MichaelRFairhurst** [comment]
unmarked as a duplicate of #38469.

#38469 will make this valid syntax, but ideally, we simply know not to insert a ! in this case at all, and instead just cast to a non-null type. This is possible if we track postdominated `DecoratedType`s instead of postdominated `Element`s.

**Comment by stereotype441** [comment]
As of ?, the migration tool produces:
```dart
class A {
  num? get foo => null;
}

void test() {
  A a = A();
  var x = (a.foo as int?)!;
  x.isEven;
}
```
Which is statically valid and has the intended semantics.  So things are much improved.  But Mike's comment (https://github.com/dart-lang/sdk/issues/39555#issuecomment-574443191) still stands: it would be better to do `a.foo as int` rather than `(a.foo as int?)!`, so I'm leaving this issue open.

**Comment by stereotype441** [comment]
As of 1c7fe7196dd7a542f7f91cd6ea2ac81e1a13146c, the null safety migration tool has been removed from active development and retired. No further work on the tool is planned.

If you still need help, or you believe this issue has been closed in error, please feel free to reopen.
